ddl/db_test.go

```diff
@@ -1846,6 +1846,109 @@ func (s *testDBSuite2) TestTableForeignKey(c *C) {
 	s.tk.MustExec("drop table if exists t1,t2,t3;")
 }
 
+func (s *testDBSuite2) TestFKOnGeneratedColumns(c *C) {
+	s.tk = testkit.NewTestKit(c, s.store)
+	s.tk.MustExec("use test")
+	// test add foreign key to generated column
+
+	// foreign key constraint cannot be defined on a virtual generated column.
+	s.tk.MustExec("create table t1 (a int primary key);")
+	s.tk.MustGetErrCode("create table t2 (a int, b int as (a+1) virtual, foreign key (b) references t1(a));", tmysql.ErrCannotAddForeign)
+	s.tk.MustExec("create table t2 (a int, b int generated always as (a+1) virtual);")
+	s.tk.MustGetErrCode("alter table t2 add foreign key (b) references t1(a);", tmysql.ErrCannotAddForeign)
+	s.tk.MustExec("drop table t1, t2;")
+
+	// foreign key constraint can be defined on a stored generated column.
```

Comment:
How about `s\defined on\reference` ? 
So as L1884,

Reply:
https://github.com/pingcap/tidb/pull/12060#discussion_r325083813 PTAL

Reply:
It seems that MySQL doc uses `reference to`，anyway, it doesn't matter.

Reply:
The `reference` is more precise in the foreign constraint scenario.

Reply:
@lonng @wjhuang2016 I think the word `reference` is not applicable in this test case, because the referred key is `a int primary key`, which is not a stored generated column.